Intersection improvements

jagill opened the pull request:
This is a WIP; I'm very interested in feedback on how to do this correctly in Rust :).

1. I've added an `RTreeObject.intersects(other: RTreeObject)` method, which checks if this obj's envelope intersects other's envelope.  This is syntactic sugar to make a common operation more readable.  **However**, it's not as useful as I hoped, because although RTreeNode implements RTreeObject, Leaf and Parent don't directly, so any intersection involving them has to write out the logic explicitly.  There might be a cool way to push down the derived `intersects` method to the enum members, but my explorations didn't find one :(.  Options that I see include:
  A. Leave it as it is,
  B. Implement `intersects` (or `intersects_envelope`) on Leaf and Parent, or
  C. Refactor to have an `Enveloped` trait that requires an `envelope()` method, and has a default implementation of `intersects_envelope()` that does the "right" thing.  I started to do this refactor, but it was getting involved and so I stopped shaving yaks.  I think there might be some good discussion about whether this would be useful, and if so, the right way to do it.

2. I've refactored the IntersectionIterator to enforce the invariant condition that the todo stack only includes items whose envelopes intersect.  While this implementation works (up to tests), it's not as slick as I wanted, largely because I couldn't make a `&RTreeNode` out of a `&ParentNode`.  If the ParentNode wasn't a reference, I could just make `RTreeNode::Parent(parent_node)`, but I couldn't get the ownership to work out given the reference returned by `RTree.root()`.  There's likely a way to make it work, if my mental model was better!

3. I also pruned some overly eager autocompletion.

Stoeoef commented on rstar/src/object.rs:
```diff
@@ -82,6 +82,11 @@ pub trait RTreeObject {
     ///
     /// Usually, this will return the object's [axis aligned bounding box](struct.AABB.html).
     fn envelope(&self) -> Self::Envelope;
+
```
In regards to a "good way" to define this intersection method: The "rust" solution to write a generic intersects method that works with multiple types would be to create a trait (e.g., `HasEnvelope`):
```
trait HasEnvelope {
  type Envelope: Envelope,
  fn envelope(&self) -> Self::Envelope;
}

impl <T: RTreeObject> HasEnvelope for T { ... }
impl <E: Envelope> HasEnvelope for E { ... } 

fn intersects<A: HasEnvelope, B: HasEnvelope>(a: &A, b: &B) -> bool {
  a.envelope().intersects(b.envelope())
}
```
With this trait, the generic intersects method should work. It is also possible to "hijack" the RTreeObject trait for this purpose, but that trait should only be implemented by objects that are meant to be inserted into RTrees.

However, for me this seems to be way too much code to maintain and understand for only a nice short cut, thus, I would prefer to have only the `Fn(RTreeObject, RTreeObject) -> bool` function and to write out any other intersections.

Stoeoef commented:
Thank you very much! This already looks promising, the only change I would make is to move the new intersect method out of `RTreeObject`.
I also agree with all the points you mention and had similar troubles when implementing the current version - like not being able to create a `RTreeNode` from a `&ParentNode`.
In regards to Point 1, I would prefer to leave this as is - the clean solution would be solution 3, but it seems to be over engineering for me.

jagill commented:
Thanks for the pointers!

1. I extracted the `intersects()` method, but it was only used in once place, so I just removed it altogether and kept the envelope intersection logic explicit.

2. I renamed `push` to `push_if_intersecting`.

3. I tried the closure, but I don't think it was clearer, and there started to be some ownership subtleties, so I left the duplicated children logic as is.

Stoeoef commented:
Thank you for the quick fixes! This looks good to me.
